pkg/messaging/nats/pubsub.go

```diff
@@ -69,20 +70,33 @@ func (ps *pubsub) Subscribe(id, topic string, handler messaging.MessageHandler)
 	if topic == "" {
 		return ErrEmptyTopic
 	}
+
 	ps.mu.Lock()
-	defer ps.mu.Unlock()
 	// Check topic
 	s, ok := ps.subscriptions[topic]
 	switch ok {
 	case true:
 		// Check topic ID
 		if _, ok := s[id]; ok {
-			return ErrAlreadySubscribed
+			// Unlocking, so that Unsubscribe() can access ps.subscriptions
+			ps.mu.Unlock()
+			if err := ps.Unsubscribe(id, topic); err != nil {
+				return err
+			}
+
+			ps.mu.Lock()
+			s, ok = ps.subscriptions[topic]
+			if !ok {
+				s = make(map[string]subscription)
+				ps.subscriptions[topic] = s
```

Comment:
Can this be contracted into:
```
ps.subscriptions[topic] = make(map[string]subscription)
```

If yes, do we need at all value `s` on line `88`? It looks like it is not used.

Reply:
Yes, this looks better to me

Reply:
Even more, I would embed check in `if` initialization - no need for two lines.

Reply:
just a min

Reply:
This reassigning of the `s` is necessary here because `ps.subscriptions[topic]` can change as soon as the lock is released. I don't only think of `Unsubscribe` call that will certainly change it, but also possible `Subscribe` calls that wait for the lock to be released - we can't know.

Reply:
@dborovcanin the way it is implemented now (in the newest change that @AryanGodara added) looks very nice to me.

Can you please review and approve if everything is OK now.